### ISSUE TITLE
OBB: Analysis of Rotation Object Detection Angles

### ISSUE THREAD

**Issue by 1chengz** [ultralytics/ultralytics#24481]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

First, I want to confirm whether the obb model exports to the ONNX format in the order of x, y, w, h, class id, confidence, angle.
Secondly, I want to ask whether the angles have a range of values, and whether different YOLO models have differences.
I found that the range of yolo26 at the official source is 0-90°, but some references have other ranges.

<img width="1736" height="244" alt="Image" src="https://github.com/user-attachments/assets/a0d90328-d6be-4aec-a245-c50ea1db580f" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @1chengz, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue routed quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this is a question about OBB export/output interpretation, please provide as much context as possible to help the team review it efficiently, such as:
- your `ultralytics` version
- the exact model used
- the export command or script
- the ONNX output shape/order you are observing
- a small reproducible example if possible
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

> First, I want to confirm whether the obb model exports to the ONNX format in the order of x, y, w, h, class id, confidence, angle.

For YOLO26-OBB, it's `[x1, y1, x2, y2, confidence, class_id, angle]`. The output is described in the docs:

https://docs.ultralytics.com/guides/end2end-detection

> Secondly, I want to ask whether the angles have a range of values, and whether different YOLO models have differences.
> I found that the range of yolo26 at the official source is 0-90°, but some references have other ranges.

All YOLO-OBB models have the range 0-90.

**Comment by raimbekovm** [comment]
hey @1chengz, both answers depend on which OBB head you exported. quick clarification with code refs and shapes from an actual export.

**1. ONNX output layout**

- **yolov8-obb and yolo11-obb** (non-end2end `OBB` head): raw output shape is `(B, 4 + nc + 1, num_anchors)`. the shape depends on `nc`: pretrained DOTAv1 weights (`nc=15`) give `(1, 20, 8400)` at 640 input; an 80-class export gives `(1, 85, 8400)`. last-dim layout is `[xc, yc, w, h, cls_prob_0, ..., cls_prob_{nc-1}, angle]`. there is no flat `class_id` or `confidence` column. you compute `class_id = argmax(cls_probs)` and `confidence = max(cls_probs)` yourself before running NMS. code: `ultralytics/nn/modules/head.py` L466-471, postprocess docstring L495-496.

- **yolo26-obb** (end2end `OBB26` head): raw output shape is `(B, max_det, 7)`, default `(1, 300, 7)`. last-dim layout is `[xc, yc, w, h, max_class_prob, class_index, angle]`. so it's confidence first, then class id, not the other way around. `class_index` is stored as a float but the values are integers in `[0, nc-1]`. code: `head.py` L491-506.

reproducible:

```python
import onnxruntime as ort, numpy as np
from ultralytics import YOLO

YOLO("yolov8n-obb.yaml").export(format="onnx", imgsz=640)
YOLO("yolo26n-obb.yaml").export(format="onnx", imgsz=640)

x = np.random.randn(1, 3, 640, 640).astype(np.float32)
print(ort.InferenceSession("yolov8n-obb.onnx").run(None, {"images": x})[0].shape)  # (1, 85, 8400)
print(ort.InferenceSession("yolo26n-obb.onnx").run(None, {"images": x})[0].shape)  # (1, 300, 7)
```

**2. angle range**

angles are always in radians. the canonical range in the codebase is `[-π/4, 3π/4)`, set by `ops.xyxyxyxy2xywhr` when ground-truth labels get converted to `xywhr` (see `ultralytics/utils/ops.py` L345-373; the docstrings of `xyxyxyxy2xywhr` and `xywhr2xyxyxyxy` both state "rotation values are in radians from [-pi/4, 3pi/4)").

at the head level the two heads differ:

- `OBB` (v8 / v11): `angle = (sigmoid(raw) - 0.25) * π`, so the model output is hard-bounded to `[-π/4, 3π/4]` (`head.py` L483). easy to verify by forcing the angle-head bias large positive / large negative: you get exactly `2.3562` and `-0.7854`.

- `OBB26` (yolo26): no sigmoid, the raw value is written out as-is (`head.py` L533-543). nothing in the head clamps it; bounding comes from training targets being encoded in `[-π/4, 3π/4)` via `xyxyxyxy2xywhr`.

so when reading the ONNX output of an OBB model, treat the angle channel as radians in `[-π/4, 3π/4)` for v8 / v11 (hard bound) and as a raw value for yolo26 (typically near `[-π/4, 3π/4)` after training). other references using `[0, π/2)` are describing a regularized DOTA-style convention where `w >= h` is enforced; that mapping is not applied at inference in this repo (`regularize_rboxes` in `ops.py` is defined but not invoked).

**Comment by 1chengz** [comment]
So, in rotated object detection, the angle is measured relative to which edge? Is it measured relative to the positive direction of the x-axis?

**Comment by 1chengz** [comment]
The rotated object detection calculates the angle with the long side of the box, right?
The radian detection range of all OBB models is within [-π/4, 3π/4), [0, π/2) is a convention in DOTA, not a mandatory specification in YOLO.

**Comment by glenn-jocher** [comment]
Yes — in Ultralytics OBB, `angle` is measured from the positive x-axis to the `w` edge in `xywhr`; during `xyxyxyxy2xywhr`, `w` and `h` are swapped when needed so `w` becomes the long side, which is why it usually matches the long-edge orientation. So your summary is correct: the repo’s canonical range is `[-pi/4, 3pi/4)`, while `[0, pi/2)` is the regularized DOTA-style convention described in the [OBB conversion helpers](https://docs.ultralytics.com/reference/utils/ops).

**Comment by raimbekovm** [comment]
to close the loop on the `0–90°` wording: that came from the OBB docs note, which was inaccurate. the canonical range is `[-π/4, 3π/4)` (`[-45°, 135°)`), and the heads are bounded to it (`OBB`: `(sigmoid(raw)-0.25)·π`; `OBB26`: raw, trained on `[-π/4, 3π/4)` targets). `[0°, 90°)` is the regularized DOTA-style form from `regularize_rboxes`, which has no callers in the codebase, so it is not applied at training or inference, and `result.obb.xywhr` is not regularized. docs fixed in #24534.

**Comment by 1chengz** [comment]
So after the model predicts, when outputting according to x, y, w, h, max_class_prob, class id, and angle, is it also assumed by default that w is greater than h?

**Comment by raimbekovm** [comment]
no, `w >= h` is not guaranteed on the model output. it is a training-time convention enforced on the labels by `xyxyxyxy2xywhr` (`ops.py` L365-367 swaps so that `w` is the long side), so the targets are always `w >= h`. but nothing in the heads or post-processing swaps `w/h` at inference, so raw predictions can come out with `w < h` — more so for `OBB26` (yolo26), which writes the raw angle channel with no clamp (`head.py` L533-543), and on out-of-distribution images. v8/v11-obb follow the convention much more tightly on in-distribution data thanks to the `(sigmoid(raw)-0.25)·π` angle clamp in `OBB`. if you need long-side form downstream, swap manually:

```python
mask = w < h
w, h = np.where(mask, h, w), np.where(mask, w, h)
theta = np.where(mask, theta + np.pi / 2, theta)
```

note `regularize_rboxes` (`ops.py` L604) is not the same operation — it normalizes the angle into `[0, π/2)`, it does not directly enforce `w >= h`, and it has no callers, so `result.obb.xywhr` and the raw ONNX output are not regularized.

**Comment by 1chengz** [comment]
The only guarantee is that the angle is the angle between w and the positive direction of the x-axis.
Currently, the angle range output by OpenCV's cv2.minAreaRect function is (0, 90], which represents the angle of clockwise rotation from the x-axis to the first encountered rectangle boundary. If you swap w and h accordingly, the range is also within [0, π). So is it still necessary to control it within [-π/4, 3π/4)? Why is it controlled within this range?

**Comment by raimbekovm** [comment]
right - the long-side / x-axis relationship is the only guaranteed semantic property. on the choice of [-π/4, 3π/4):

any 180° window is mathematically equivalent for an OBB with w >= h (rotating by π gives the same rectangle), so the question is which 180° to pick. the shift by -π/4 has two practical reasons.

first, axis-aligned objects sit deep in the interior: horizontal (θ=0) and vertical (θ=π/2) are each π/4 from the nearest boundary. with [0, π) horizontal lands exactly at the edge, so slightly-rotated objects regress from `π-ε` vs `+ε` - a wrap discontinuity right where most real objects sit. with [-π/4, 3π/4) the wrap lands at ±45° / 135°, where w ≈ h is geometrically ambiguous anyway.

second, the v8 / v11 `OBB` head encodes the angle as `(sigmoid(raw) - 0.25) * π` (`head.py:483`), a smooth hard-bounded mapping into exactly [-π/4, 3π/4] with raw=0 landing at the midpoint π/4. [0, π) would force either a modulo wrap or a wider stretch with worse edge calibration. yolo26's `OBB26` writes the angle raw (`head.py:533-543`), but labels are still encoded in [-π/4, 3π/4) by `xyxyxyxy2xywhr` (`ops.py:365-373`), so the head learns the same range.

[0, π/2) (DOTA-style) is a further folding that re-enforces w >= h at every π/2 boundary; `regularize_rboxes` (`ops.py:604`) exists for that but is never called in train / val / inference / export, so `result.obb.xywhr` and the raw ONNX output stay in [-π/4, 3π/4).

**Comment by 1chengz** [comment]
It is known that the image coordinate system is different from the mathematical coordinate system, so is the range of [-π/4, 3π/4) different in the coordinate system? For example, in the image coordinate system, are the angles corresponding to the positive directions of the x-axis and y-axis positive or negative?

**Comment by raimbekovm** [comment]
the range itself (`[-π/4, 3π/4)`) is just a numerical bound on the angle channel — it does not change with coordinate system. it comes from how `xyxyxyxy2xywhr` normalizes the encoder output (`ops.py` L368-371).

what does depend on convention is the sign of θ. ultralytics uses the image coordinate system (same as opencv): origin top-left, x → right, y → down. with that:

- θ = 0 → long side parallel to the positive x-axis (horizontal on screen)
- θ = +π/2 → long side parallel to the positive y-axis (vertical, pointing down on screen)
- positive θ = clockwise rotation on screen (because y is flipped relative to the math convention; this matches `cv2.minAreaRect`)

verified empirically: a rectangle hand-built with the standard rotation matrix `[[cos,-sin],[sin,cos]]` and angle `+30°` encodes back to `θ = +30°`; the same rectangle with `-30°` encodes to `-30°`.
